pages/posts/schema-static-analysis.md

```diff
@@ -0,0 +1,167 @@
+
+What if we have a recursive schema, like a schema that validates a linked list or a binary tree?  In these cases, the recursion stops when the instance no longer has any data that needs validating, i.e. when you read the end of the list or a leaf on the node.  To handle this, we can take the same "constraint template" approach as in the previous section.
+
+_The template solution actually works for a lot of cases.  The real trick is figuring out when you need to use it._
+
+## Dynamic references
+
+Dynamic references, on the other hand, generally boil down to one thing: dynamic scope.  Dynamic scope is the ordered set of resource IDs (generally set by `$id`) that evaluation enters and leaves.  (Think of a stack, pushing when entering a resource, and popping when leaving.)  The dynamic scope is influenced by two things:
+
+- where the evaluation started
+- instance data
+
+### Root schema dynamics
+
+Last year I wrote a [post](./dynamicref-and-generics) describing how to use `$dynamicRef` to model generic types in languages.  The idea went like this:
+
+1. Start by defining a generic schema, using a `$dynamicRef` to a `$dynamicAnchor` to identify an undefined "type" parameter.
+2. Define multiple secondary schemas that reference the generic one and use their own `$dynamicAnchor` to define the "type" parameter: one for each type.
+
+Using this approach, if you start evaluation from the generic schema, evaluation will fail because the "type" isn't defined.  However, starting evaluation from a secondary schema redirects the `$dynamicRef` resolution to the one defined in the secondary schema.  This different resolution can allow an instance to pass validation.
+
+This is a great example of a dynamic reference that _can_ be resolved without an instance present.  You only need the starting point of the evaluation.  Specifically, you need to know where the dynamic scope starts in order to identify the reference target.
+
+### Data-driven dynamics
+
+Another way the dynamic scope can change is through some kind of conditional logic.  [This test](https://github.com/jdesrosiers/JSON-Schema-Test-Suite/blob/06fc43d767e11369be443a60523261a3ce032423/tests/draft2020-12/dynamicRef.json#L390-L475) from the JSON Schema Test Suite is a good example.  It uses the same ideas from from the generics post, but instead of having separate schemas, everything is bundled up into one.
+
+For this case, depending on the value of the `kindOfList` instance property, the items in the array are expected to either be numbers or strings.  Mechanically, this is determined by a set of `if`/`then`/`else` keywords that directs evaluation into either of the `numberList` or `stringList` definitions, both of which define `$dynamicAnchor: itemType` and reference into the `genericList` definition which contains `$dynamicRef: #itemType`.
```

Comment:
This explanation was really helpful in building up the mental model while looking at the test.
I wonder if it's worth mentioning in passing, that the dynamic scope is in relation to the "resource", and that the subschema which contains the dynamic anchor doesn't actually have to be used for evaluation in order for it to be referenced by a dynamic anchor. This was a critical jump in understanding for me personally. Leave it up to you.

Reply:
I think that's probably better for a deep-dive on `$dynamic*`.  I'm not sure it fits here.